components/raftstore-v2/src/raft/storage.rs

```diff
@@ -358,4 +446,78 @@ mod tests {
+        let raft_engine =
+            engine_test::raft::new_engine(&format!("{}", path.path().join("raft").display()), None)
+                .unwrap();
+        let mut wb = raft_engine.log_batch(10);
+        write_initial_states(&mut wb, region.clone()).unwrap();
+        assert!(!wb.is_empty());
+        raft_engine.consume(&mut wb, true).unwrap();
+        // building a tablet factory
+        let ops = DbOptions::default();
+        let cf_opts = ALL_CFS.iter().map(|cf| (*cf, CfOptions::new())).collect();
+        let factory = Arc::new(TestTabletFactoryV2::new(
+            path.path().join("tablet").as_path(),
+            ops,
+            cf_opts,
+        ));
+        // create tablet with region_id 1
+        let tablet = factory
+            .open_tablet(1, Some(10), OpenOptions::default().set_create_new(true))
+            .unwrap();
+        // setup read runner worker and peer storage
+        let mut worker = Worker::new("test-read-worker").lazy_build("test-read-worker");
+        let sched = worker.scheduler();
+        let logger = slog_global::borrow_global().new(o!());
+        let mut s = Storage::uninit(
+            6,
+            region.clone(),
+            raft_engine.clone(),
+            sched.clone(),
+            &logger.clone(),
+        )
+        .unwrap();
+        let (router, rx) = TestRouter::new();
+        worker.start(ReadRunner::new(router.clone(), raft_engine));
+        // setup peer applyer
+        let mut apply = Apply::new(
+            region.get_peers()[0].clone(),
+            RegionLocalState::default(),
+            router,
+            CachedTablet::new(Some(tablet)),
+            sched,
+            logger,
+        );
+
+        // test get snapshot
+        let snap = s.snapshot(0, 0);
+        let unavailable = RaftError::Store(StorageError::SnapshotTemporarilyUnavailable);
+        assert_eq!(snap.unwrap_err(), unavailable);
+        let gen_task = s.gen_snap_task.borrow_mut().take().unwrap();
+        apply.schedule_gen_snapshot(gen_task);
+        let res = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+        s.on_snapshot_generated(res);
+        let snap = match *s.snap_state.borrow() {
+            SnapState::Generated(ref snap) => *snap.clone(),
+            ref s => panic!("unexpected state: {:?}", s),
+        };
+        assert_eq!(snap.get_metadata().get_index(), 0);
+        assert_eq!(snap.get_metadata().get_term(), 0);
+        assert!(snap.get_data().is_empty());
+
+        // test cancel snapshot
+        let snap = s.snapshot(0, 0);
+        let unavailable = RaftError::Store(StorageError::SnapshotTemporarilyUnavailable);
+        assert_eq!(snap.unwrap_err(), unavailable);
+        let gen_task = s.gen_snap_task.borrow_mut().take().unwrap();
+        apply.schedule_gen_snapshot(gen_task);
+        let res = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+        s.cancel_generating_snap(None);
+        assert_eq!(*s.snap_state.borrow(), SnapState::Relax);
```

Comment:
How about also adding cases mentioned in this PR?

Reply:
I added TODO here. will add the test after implement async task works, otherwise the snapshot metadata is a mock data.